### PR TITLE
feat: Improve LSP console (#867)

### LINKED ISSUE #867: Improve LSP console
Improve LSP console with

* format LSP request responses notification with the same format than vscode
* when langauge server tree node is selected display on the right the description of the ls and the trace settings. It will avoid opening language server to configure trace level and it will give some to the user some understand how to activate trace fir a given ls
* provide a popup menu on the process tree node to stop and restart the ls
* display pid on the process tree node

### PR DESCRIPTION
**Pull request by angelozerr**
feat: Improve LSP console (#867)

Fixes #867

### CODE REVIEW

**Comment by angelozerr** [comment]
Now the LSP console display the pid:

image

**Comment by angelozerr** [comment]
Now with contextual menu, you can stop language server

image

and restart language server:

image

**Comment by angelozerr** [comment]
In Language Server prferences, server icon are now used:

image

**Comment by angelozerr** [comment]
Now when you click on language server node, you can see description and trace settings:

image

Please note that if you update trace settings from Language Servers settings, the combo is noy updated. We need to manage settings listener to support that. I would like to do that in a separate PR.

File: src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/actions/RestartServerAction.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    public static final String ACTION_ID = "com.redhat.devtools.intellij.lsp4ij.console.actions.RestartServerAction";
+
+    public RestartServerAction() {
+        super(LanguageServerBundle.message("language.server.actions.restart"));
+    }
+
+    @Override
+    public void actionPerformed(@NotNull AnActionEvent e) {
+        LanguageServerWrapper ls = (LanguageServerWrapper) e.getDataContext().getData(TreeDataProvider.LANGUAGE_SERVER_DATA_ID);
+        try {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/actions/RestartServerAction.java:23]:
```java
Object lsData = e.getDataContext().getData(TreeDataProvider.LANGUAGE_SERVER_DATA_ID);
if (lsData instanceof LanguageServerWrapper) {
 try {
    ((LanguageServerWrapper)ls).start();
 } catch (IOException e) {
  LOGGER.error("Failed restarting server", e);
 }
}
```

File: src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/actions/StopServerAction.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    public static final String ACTION_ID = "com.redhat.devtools.intellij.lsp4ij.console.actions.StopServerAction";
+
+    public StopServerAction() {
+        super(LanguageServerBundle.message("language.server.actions.stop"));
+    }
+
+    @Override
+    public void actionPerformed(@NotNull AnActionEvent e) {
+        LanguageServerWrapper ls = (LanguageServerWrapper) e.getDataContext().getData(TreeDataProvider.LANGUAGE_SERVER_DATA_ID);
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/actions/StopServerAction.java:20]: cf previous comment

File: src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/explorer/TracingMessageConsumer.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        if (messageConsumer instanceof StreamMessageConsumer) {
+            return consumeMessageSending(message, now, date, serverTrace);
+        } else if (messageConsumer instanceof RemoteEndpoint) {
+            return consumeMessageReceiving(message, now, date, serverTrace);
+        } else {
+            return String.format("Unknown MessageConsumer type: %s", messageConsumer);
+        }
+    }
+
+    private String consumeMessageSending(Message message, Instant now, String date, ServerTrace serverTrace) {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/explorer/TracingMessageConsumer.java:71]: Maybe the different formats can be stored as private static final constants

File: src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/explorer/TreeDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package com.redhat.devtools.intellij.quarkus.lsp4ij.console.explorer;
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/explorer/TreeDataProvider.java:1]: missing header

**Review by fbricon — CHANGES_REQUESTED** [review]
Overall works fine, only got a couple nitpicks

File: src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/LanguageServerWrapper.java [hunk 5]
```diff
@@ -496,18 +471,18 @@ synchronized void stop() {
                 } catch (InterruptedException ex) {
                     Thread.currentThread().interrupt();
                 } catch (Exception ex) {
-                    LOGGER.error("Error while shutdown the language server '" + serverDefinition.id + "'", ex);
+                    LOGGER.warn("Error while shutdown the language server '" + serverDefinition.id + "'", ex);
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/LanguageServerWrapper.java:474]: I remember that I do that to log TomeoutException as warning. I think we should warn TimeOutException and log as error the other error kind. What do you think about that?

[the remark this answers is not in the document]
Comment (fbricon) [reply src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/LanguageServerWrapper.java:474]: fine by me

[the remark this answers is not in the document]
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/LanguageServerWrapper.java]: Ok fixed

[the remark this answers is not in the document]
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/LanguageServerWrapper.java]: fixed

File: src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/actions/StopServerAction.java (hunk 2 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/actions/StopServerAction.java]: fixed

File: src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/explorer/TracingMessageConsumer.java (hunk 3 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/explorer/TracingMessageConsumer.java:71]: I would like to keep this code to keep the maximum similarity from the origin code https://github.com/eclipse-lsp4j/lsp4j/blob/main/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/TracingMessageConsumer.java

File: src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/explorer/TreeDataProvider.java (hunk 4 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/console/explorer/TreeDataProvider.java]: fixed

File: src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/LanguageServerWrapper.java [hunk 6]
```diff
@@ -496,18 +471,18 @@ synchronized void stop() {
...
 
-            if (serverFuture != null) {
-                serverFuture.cancel(true);
-            }
-
             if (languageServerInstance != null) {
                 languageServerInstance.exit();
             }
 
+            if (serverFuture != null) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/quarkus/lsp4ij/LanguageServerWrapper.java:486]: I add a comment to explain why.

**Comment by angelozerr** [comment]
Thanks so much @fbricon for your great review!

**Pull request merged**